core/dbt/tests/util.py

```diff
@@ -311,17 +375,47 @@ def generate_update_clause(adapter, clause) -> str:
             raise TestProcessingException("Invalid update_rows clause: no src_col")
         add_to = clause["src_col"]
         kwargs = {k: v for k, v in clause.items() if k in ("interval", "number")}
-        with patch.object(providers, "get_adapter", return_value=adapter):
-            with adapter.connection_named("_test"):
-                return adapter.timestamp_add_sql(add_to=add_to, **kwargs)
+        with get_connection(adapter):
+            return adapter.timestamp_add_sql(add_to=add_to, **kwargs)
     elif clause_type == "add_string":
         for key in ["src_col", "value"]:
             if key not in clause:
                 raise TestProcessingException(f"Invalid update_rows clause: no {key}")
         src_col = clause["src_col"]
         value = clause["value"]
         location = clause.get("location", "append")
-        with patch.object(providers, "get_adapter", return_value=adapter):
-            with adapter.connection_named("_test"):
-                return adapter.string_add_sql(src_col, value, location)
+        with get_connection(adapter):
+            return adapter.string_add_sql(src_col, value, location)
     return ""
+
+
+@contextmanager
+def get_connection(adapter, name="_test"):
+    # Since the 'adapter' in dbt.adapters.factory may have been replaced by execution
+    # of dbt commands since the test 'adapter' was created, we patch the 'get_adapter' call in
+    # dbt.context.providers, so that macros that are called refer to this test adapter.
+    # This allows tests to run normal adapter macros as if reset_adapters() were not
+    # called by handle_and_check (for asserts, etc).
+    with patch.object(providers, "get_adapter", return_value=adapter):
+        with adapter.connection_named(name):
+            conn = adapter.connections.get_thread_connection()
+            yield conn
```

Comment:
I wonder if there's a way to use the current adapter instead.  This method was copied from test/integration/base.py. It would be nice to not have to do this.

Reply:
Yay! Was able to take off the patch bit.

Reply:
I removed a permission test after this change because it was breaking. The test was flaky to start with and really wasn't doing the right thing anyway. I figured that we have lots of tests that show that users with the right permissions work :) but if you think we still need something similar, we should make a separate ticket and do it right (creating randomized user name, no permissions, test, change permissions, test, drop user and permissions).

Reply:
Oh, just fyi, the issue was that after running the permission SQL in the dropped test, the "drop schema" in the project fixture was failing because it didn't have the right permissions.